ci: use circleci windows preview

filipesilva opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [x] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the new behavior?
Added a CircleCI Windows track, similar to https://github.com/angular/angular-cli/pull/14687.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

filipesilva commented on .circleci/bazel_win.rc:
```diff
@@ -0,0 +1,30 @@
+# TODO(gmagolan): Hide progress again once build performance improves
+# Presently, CircleCI can timeout during bazel test ... with the following
+# error: Too long with no output (exceeded 10m0s)
+# build --noshow_progress
```
This needs to be uncommented.

filipesilva commented on .circleci/bazel_win.rc:
```diff
@@ -0,0 +1,30 @@
+# Bazel doesn't calculate the memory ceiling correctly when running under Docker.
+# Limit Bazel to consuming resources that fit in CircleCI "xlarge" class
+# https://circleci.com/docs/2.0/configuration-reference/#resource_class
+# build --local_resources=14336,8.0,1.0
```
Need to figure out the resources on the CircleCI windows machines to set here.

filipesilva replied:
It's 15gigs ram: https://github.com/CircleCI-Public/windows-preview-docs#frequently-asked-questions

filipesilva commented on .circleci/bazel_win.rc:
```diff
@@ -0,0 +1,30 @@
+
+# Save downloaded repositories in a location that can be cached by CircleCI. This helps us
+# speeding up the analysis time significantly with Bazel managed node dependencies on the CI.
+# build --repository_cache=/home/circleci/bazel_repository_cache
```
File access on CircleCI windows machines is very slow atm so caching things is a net loss right now. I should add context here explaining this.

filipesilva replied:
Updated

filipesilva commented on .circleci/config.yml:
```diff
@@ -667,93 +688,125 @@ jobs:
+        #   branches:
+        #     ignore:
+        #       - /pull\/.*/
+      # - setup
```
Just ignoring all other jobs for now to test the windows one.

filipesilva commented on .circleci/config.yml:
```diff
@@ -667,93 +688,125 @@ jobs:
+        # filters:
+        #   branches:
+        #     ignore:
+        #       - /pull\/.*/
```
Need to add context about why this job is master only atm.

filipesilva commented on .circleci/windows-env.ps1:
```diff
@@ -0,0 +1,22 @@
+# Use lf as the line ending. The //packages/bazel/test/ng_package:example_package test doesn't
+# deal well with windows line endings. We should fix the test instead really.
+git config --global core.autocrlf input
+git config --global core.eol lf
```
No need for these. I can remove them.

filipesilva commented on package.json:
```diff
@@ -7,7 +7,7 @@
   "license": "MIT",
   "engines": {
-    "node": ">=10.9.0 <11.0.0",
+    "node": ">=10.9.0 <13.0.0",
```
Updating to Node 12 because it is what's easier to install via Chocolatey.

filipesilva replied:
Done, and added a comment for this to happen in the future.

filipesilva commented on package.json:
```diff
@@ -26,7 +26,9 @@
     "list-fixme-ivy-targets": "bazel query --output=label 'attr(\"tags\", \"\\[.*fixme-ivy.*\\]\", //...) except kind(\"sh_binary\", //...) except kind(\"devmode_js_sources\", //...)' | sort",
-    "bazel": "bazel"
+    "bazel": "bazel",
+    "circleci-win-ve": "yarn bazel test --build_tag_filters=-ivy-only --test_tag_filters=-ivy-only,-browser:chromium-local //...",
```
These were added because of the situation described in the config.yml comment:
```
      # Ran into a some command parsing problem where `-browser:chromium-local` was converted to
      # `-browser: chromium-local` (a space was added) in https://circleci.com/gh/angular/angular/357511.
      # Probably a powershell command parsing thing. This way there's no problem.
      - run: yarn circleci-win-ve
      - run: yarn circleci-win-ivy
```

filipesilva replied:
Added comment here instead.

filipesilva commented on packages/bazel/test/ng_package/example_package.spec.ts:
```diff
@@ -92,7 +92,7 @@ function runPackageGoldTest(testPackage: TestPackage) {
 
   // Gold file content from source control. We expect that the output of the package matches this.
-  const expected = fs.readFileSync(goldenFilePath, 'utf-8');
+  const expected = readFileContents(goldenFilePath);
```
Golden files should be read like to to avoid line ending differences between OSs.

filipesilva replied:
Done

filipesilva commented on packages/compiler-cli/ngcc/test/BUILD.bazel:
```diff
@@ -58,6 +58,7 @@ ts_library(
 
 jasmine_node_test(
     name = "integration",
+    timeout = "long",  # 900 seconds
```
This test takes ages. It needs a bigger timeout. Other similar tests have this timeout too.

gkalpak reviewed (commented):
Looks promising (except that CI fails on `yarn bazel info` :grin:)

josephperrott reviewed (commented):
This might be a naive thought, but are we able to take advantage of bash on windows to reuse the scripting we already have?

filipesilva commented:
@josephperrott the CircleCI windows machines do have the [GitBash shell available](https://github.com/CircleCI-Public/windows-preview-docs#shells). 

But in this specific case we really should not use it because Bazel must not be ran from Git Bash because of the problems discussed in these issues:
- https://github.com/bazelbuild/bazel/issues/5751
- https://github.com/bazelbuild/bazel/issues/5724#issuecomment-410194038
- https://github.com/bazelbuild/bazel/issues/6339#issuecomment-441600879

It's an especially bad problem actually, because it can alternate between working and not working in gitbash depending on what the status of your local cache is. If you a full cache you might not hit those DLL errors because no bash binary needs to be used.

I should add this as context to the windows shell area of the circleci config.

@gkalpak just rebased and repushed, and it doesn't seem to be failing there anymore.

filipesilva commented on .circleci/config.yml:
```diff
@@ -667,93 +688,125 @@ jobs:
+    steps:
+      - checkout
+      - *env_win
+      - *post_checkout_win
```
Kinda depends on what jobs we want to run. I don't think we'll want to run all jobs. 

In my head the logic is that tooling that needs to run on Windows should be tested. These are things like Angular CLI, @angular/compiler-cli, and local Bazel packages. But then stuff like AIO and most of the FW packages proper don't really need to be tested on Windows.

I'm not really following this principle on the PR though. It's something we should discuss.

filipesilva commented on .circleci/bazel_win.rc:
```diff
@@ -0,0 +1,30 @@
+# These options are enabled when running on CI
+# We do this by copying this file to /etc/bazel.bazelrc at the start of the build.
```
Done

filipesilva commented:
Failing on an Ivy test:
```
Failures:
1) nodejs file system watcher related methods test fs.watchFile has been patched as eventTask
  Message:
    Expected 0 to be 16.
  Stack:
    Error: Expected 0 to be 16.
        at <Jasmine>
        at StatWatcher.<anonymous> (C:\users\circleci\_bazel_circleci\u4uoan2j\execroot\angular\bazel-out\x64_windows-fastbuild\bin\packages\zone.js\test\node\fs.spec.js:75:47)
        at ZoneDelegate.invokeTask (C:\users\circleci\_bazel_circleci\u4uoan2j\execroot\angular\bazel-out\x64_windows-fastbuild\bin\packages\zone.js\lib\zone.js:409:31)
        at ProxyZoneSpec.onInvokeTask (C:\users\circleci\_bazel_circleci\u4uoan2j\execroot\angular\bazel-out\x64_windows-fastbuild\bin\packages\zone.js\lib\zone-spec\proxy.js:139:39)
        at ZoneDelegate.invokeTask (C:\users\circleci\_bazel_circleci\u4uoan2j\execroot\angular\bazel-out\x64_windows-fastbuild\bin\packages\zone.js\lib\zone.js:408:36)
        at Object.onInvokeTask (C:\users\circleci\_bazel_circleci\u4uoan2j\execroot\angular\bazel-out\x64_windows-fastbuild\bin\packages\zone.js\lib\zone.js:284:29)
        at ZoneDelegate.invokeTask (C:\users\circleci\_bazel_circleci\u4uoan2j\execroot\angular\bazel-out\x64_windows-fastbuild\bin\packages\zone.js\lib\zone.js:408:36)
        at Zone.runTask (C:\users\circleci\_bazel_circleci\u4uoan2j\execroot\angular\bazel-out\x64_windows-fastbuild\bin\packages\zone.js\lib\zone.js:177:47)
        at ZoneTask.invokeTask (C:\users\circleci\_bazel_circleci\u4uoan2j\execroot\angular\bazel-out\x64_windows-fastbuild\bin\packages\zone.js\lib\zone.js:484:34)
        at StatWatcher.ZoneTask.invoke (C:\users\circleci\_bazel_circleci\u4uoan2j\execroot\angular\bazel-out\x64_windows-fastbuild\bin\packages\zone.js\lib\zone.js:473:48)
        at StatWatcher.emit (events.js:198:13)
        at StatWatcher.EventEmitter.emit (domain.js:448:20)
        at StatWatcher.onchange (internal/fs/watchers.js:50:8)
  Message:
    Expected 0 to be 12.
  Stack:
    Error: Expected 0 to be 12.
        at <Jasmine>
        at StatWatcher.<anonymous> (C:\users\circleci\_bazel_circleci\u4uoan2j\execroot\angular\bazel-out\x64_windows-fastbuild\bin\packages\zone.js\test\node\fs.spec.js:76:47)
        at ZoneDelegate.invokeTask (C:\users\circleci\_bazel_circleci\u4uoan2j\execroot\angular\bazel-out\x64_windows-fastbuild\bin\packages\zone.js\lib\zone.js:409:31)
        at ProxyZoneSpec.onInvokeTask (C:\users\circleci\_bazel_circleci\u4uoan2j\execroot\angular\bazel-out\x64_windows-fastbuild\bin\packages\zone.js\lib\zone-spec\proxy.js:139:39)
        at ZoneDelegate.invokeTask (C:\users\circleci\_bazel_circleci\u4uoan2j\execroot\angular\bazel-out\x64_windows-fastbuild\bin\packages\zone.js\lib\zone.js:408:36)
        at Object.onInvokeTask (C:\users\circleci\_bazel_circleci\u4uoan2j\execroot\angular\bazel-out\x64_windows-fastbuild\bin\packages\zone.js\lib\zone.js:284:29)
        at ZoneDelegate.invokeTask (C:\users\circleci\_bazel_circleci\u4uoan2j\execroot\angular\bazel-out\x64_windows-fastbuild\bin\packages\zone.js\lib\zone.js:408:36)
        at Zone.runTask (C:\users\circleci\_bazel_circleci\u4uoan2j\execroot\angular\bazel-out\x64_windows-fastbuild\bin\packages\zone.js\lib\zone.js:177:47)
        at ZoneTask.invokeTask (C:\users\circleci\_bazel_circleci\u4uoan2j\execroot\angular\bazel-out\x64_windows-fastbuild\bin\packages\zone.js\lib\zone.js:484:34)
        at StatWatcher.ZoneTask.invoke (C:\users\circleci\_bazel_circleci\u4uoan2j\execroot\angular\bazel-out\x64_windows-fastbuild\bin\packages\zone.js\lib\zone.js:473:48)
        at StatWatcher.emit (events.js:198:13)
        at StatWatcher.EventEmitter.emit (domain.js:448:20)
        at StatWatcher.onchange (internal/fs/watchers.js:50:8)
```

filipesilva commented:
Now the same test that failed above, was just flaky 🤷‍♂ 
```
//packages/zone.js/test:test_node                                         FLAKY, failed in 1 out of 2 in 17.2s
```

filipesilva commented:
@JiaLiPassion do you know anything about that zone.js test being flaky on Windows?

JiaLiPassion commented:
@filipesilva,  I will check it in windows env, thanks.

JiaLiPassion commented:
@filipesilva , I tried to run `yarn bazel run packages/zone.js/test:test_node` on my windows environment about 10 times, all success. Not sure what is the problem, the environment is 
1. Windows 10
2. nodejs 10.16

filipesilva commented:
@JiaLiPassion we've decided to disable it only on the Windows CircleCI for now.

@josephperrott here is a successful run, before I added the master only filter: https://circleci.com/gh/angular/angular/379604

josephperrott commented on .circleci/bazel_win.rc:
```diff
@@ -0,0 +1,29 @@
+# speeding up the analysis time significantly with Bazel managed node dependencies on the CI.
+# TODO(filipesilva): file access on CircleCI windows machines is very slow atm so caching things is
+# a net loss right now. Uncomment this when it's faster.
+build --repository_cache=C:/Users/circleci/bazel_repository_cache
```
Is this TODO still applicable?  Or should we have this line commented still?

filipesilva replied:
We still can't cache it, but it needs to go somewhere anyway. Since there's already a TODO about circleci caching in the config, I don't think this comment is needed. Removing.

josephperrott commented on .circleci/bazel_win.rc:
```diff
@@ -0,0 +1,29 @@
+# These options are enabled when running on CI
```
Are we have to create a common `.circleci/base_bazel.rc`?

My thinking is that we can prevent duplicating the rules like

`build --noshow_progress` and `test --flaky_test_attempts=2`

filipesilva replied:
Added a common bazelrc from which both linux and windows ones import.

josephperrott commented on .circleci/bazel_win.rc:
```diff
@@ -0,0 +1,29 @@
+build --announce_rc
+
+# Workaround https://github.com/bazelbuild/bazel/issues/3645
+# Bazel doesn't calculate the memory ceiling correctly when running under Docker.
```
Are we running inside of a docker container?  I remember reading somewhere in our timeline of working on this that we are running in the VM not the docker image for windows.  Should we still be doing this?

filipesilva replied:
No, we are not. You're right, this line isn't necessary. Removing.

gkalpak commented on .circleci/config.yml:
```diff
@@ -75,6 +75,19 @@ var_6: &job_defaults
+  # GitBash. These issues discuss why:
+  # https://github.com/bazelbuild/bazel/issues/5751
+  # https://github.com/bazelbuild/bazel/issues/5724#issuecomment-410194038
+  # https://github.com/bazelbuild/bazel/issues/6339#issuecomment-441600879
```
Love this comment 💯

gkalpak commented on .circleci/config.yml:
```diff
@@ -667,6 +701,37 @@ jobs:
+      - *env_win
+      - *post_checkout_win
+      - run: node --version
+      - run: yarn --version
```
It is good to ahve those, but imo they belong in the `env_win` step.

filipesilva replied:
I agree. Moved there.

filipesilva replied:
Discussed on Slack, it's hard to setup the windows path and use things that need it in the same script so we are leaving it here for now. Maybe later we'll actually use the existing `init_environment` anchor that linux uses via the Bash shell on Windows.

gkalpak commented on .circleci/config.yml:
```diff
@@ -667,6 +701,37 @@ jobs:
+      # Probably a powershell command parsing thing. This way there's no problem.
+      - run:
+          command: yarn circleci-win-ve
+          no_output_timeout: 120m
```
😱 🙀 😱

gkalpak commented on package.json:
```diff
@@ -26,7 +27,10 @@
+    "bazel": "bazel",
+    "//circleci-win-comment": "See the test-win circleci job for why these are needed. If they are not needed anymore, remove them.",
+    "circleci-win-ve": "yarn bazel test --build_tag_filters=-ivy-only --test_tag_filters=-ivy-only,-browser:chromium-local,-fixme-windows //...",
+    "circleci-win-ivy": "yarn bazel test --define=compile=aot --build_tag_filters=-no-ivy-aot,-fixme-ivy-aot --test_tag_filters=-no-ivy-aot,-fixme-ivy-aot,-browser:chromium-local,-fixme-windows //..."
```
`yarn` shouldn't be needed (nor `"bazel": "bazel"` above, but that's a different story :grin:).

filipesilva replied:
Replaced `yarn bazel` -> `bazel` on the `circleci-win-*` scripts.

gkalpak commented on .circleci/windows-env.ps1:
```diff
@@ -0,0 +1,17 @@
+new-item -path $profile -itemtype file -force
+# Paths for nodejs, npm, yarn, and msys2. Use single quotes to prevent interpolation.
+# Add before the original path to use msys2 instead of the installed gitbash.
+Add-Content $profile '$Env:path = "C:\Program Files\nodejs\;C:\Users\circleci\AppData\Roaming\npm\;C:\Program Files (x86)\Yarn\bin\;C:\tools\msys64\usr\bin\;" + $Env:path'
```
Should `C:\Program Files` and `C:\Program Files (x86)` be `${Env:ProgramFiles}` and `${Env:ProgramFiles(x86)}` respectively?

filipesilva replied:
Yes, I think that would be better. Changed.

gkalpak reviewed (approved):
:tada:

filipesilva commented:
Sometimes I see this error on the yarn install rule https://circleci.com/gh/angular/angular/380183:
```
ERROR: An error occurred during the fetch of repository 'npm':
   yarn_install failed: $ node tools/yarn/check-yarn.js
...
[4/5] Linking dependencies...
info If you think this is a bug, please open a bug report with the information provided in "C:\\users\\circleci\\ng\\yarn-error.log".
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
 (warning " > tsickle@0.34.3" has incorrect peer dependency "typescript@~3.3.1".
error An unexpected error occurred: "EPERM: operation not permitted, unlink 'C:\\users\\circleci\\ng\\node_modules\\@bazel\\bazel-win32_x64\\bazel-0.27.0-windows-x86_64.exe'".
Process stalled
Active handles:
  - Socket
  - Socket
  - Socket
)
```

cc @alexeagle @gregmagolan

gkalpak commented on .circleci/bazel_linux.rc:
```diff
@@ -2,29 +2,16 @@
 
+# Import config items common to both Linux and Windows setups.
+# https://docs.bazel.build/versions/master/guide.html#bazelrc-syntax-and-semantics
+import %workspace%/.circleci/bazel_common.rc
```
Where does `%workspace%` come from? Is it a bazel thing?

filipesilva replied:
It's a bazel thing referenced on that link:
>To specify a path that is relative to the workspace root, write import %workspace%/path/to/bazelrc.

As far as I can tell the import semantics for relative paths mean it's relative to your CWD, not to the file. So it makes more sense to pin down the workspace proper. I found this out because I was actually copying the common file together with the OS specific file before and that didn't work as I expected.

gkalpak replied:
Doh! RTFM (esp. since the link was right above) :grin:)

gkalpak reviewed (approved):
👍 on renaming `.circleci/bazel.rc` to `.circleci/bazel_linux.rc`.
This file is referenced (in comments) in a couple more files:
- [.bazelrc][1]
- [docs/BAZEL.md][2]

[1]: https://github.com/angular/angular/blob/890c1f4652c4cd2ded718ff0e974d80c683a548f/.bazelrc#L96
[2]: https://github.com/angular/angular/blame/890c1f4652c4cd2ded718ff0e974d80c683a548f/docs/BAZEL.md#L180

filipesilva commented:
@gkalpak updated the reference in `.bazelrc` to talk about both config files by their updated names. Also updated the ref in `docs/BAZEL.md`. But I'm pretty sure that's outdated because the file doesn't include anything about proxies anymore. I don't understand the new setup and don't want to mess the instructions up though.

filipesilva commented:
Reported the windows failure on rules_nodejs https://github.com/bazelbuild/rules_nodejs/issues/894

gkalpak commented:
FWIW, #31413 uses `.bazelrc.common` as the name for the common configs :grin:

filipesilva commented:
@gkalpak I changed the names used here to `bazel.something.rc` as well to keep it consistent with #31413 but did not move them around as in that PR.

filipesilva committed:
build: add reminder to keep engines in sync

filipesilva committed:
test(bazel): use cross-platform file read of golden file

filipesilva committed:
ci: use no_output_timeout for long jobs

filipesilva committed:
ci: use circleci windows preview

gkalpak commented on .circleci/config.yml:
```diff
@@ -80,6 +94,14 @@ var_7: &post_checkout
+      if (Test-Path env:CIRCLE_PR_NUMBER) {
+        git config user.name "angular-ci"
+        git config user.email "angular-ci"
+        node tools\rebase-pr.js angular/angular-cli $env:CIRCLE_PR_NUMBER }
```
Why not have an `else` block (for consistency with the `post_checkout` step)?

filipesilva replied:
Done

gkalpak commented on .circleci/config.yml:
```diff
@@ -766,6 +847,16 @@ workflows:
       - test_zonejs:
           requires:
             - setup
+      # This job is very slow so we run it on non-PRs only for now.
```
Adding a 2.5+ hour job is not an option even for non-PR builds. I suggest we run it as a cronjob for now (maybe once per day and report failure to the `fw-caretaker` channel - as we do for other jobs). WDYT?

filipesilva replied:
It should be around 1h. Bazel jobs on both windows and linux started taking ~2x as long sometime in the last 3 weeks. See https://angular-team.slack.com/archives/CH58NF3MG/p1564066641126100 for context.

gkalpak commented on docs/BAZEL.md:
```diff
@@ -177,7 +177,7 @@ Contact Alex Eagle with questions.
 1. In `.circleci/config.yml`, each CircleCI job downloads a proxy binary, which is built from https://github.com/notnoopci/bazel-remote-proxy. The download is done by running `.circleci/setup_cache.sh`. When the feature graduates from experimental, this proxy will be installed by default on every CircleCI worker, and this step will not be needed.
 1. Next, each job runs the `setup-bazel-remote-cache` anchor. This starts up the proxy running in the background. In the CircleCI UI, you'll see this step continues running while later steps run, and you can see logging from the proxy process.
-1. Bazel must be configured to connect to the proxy on a local port. This configuration lives in `.circleci/bazel.rc` and is enabled because we overwrite the system Bazel settings in /etc/bazel.bazelrc with this file.
+1. Bazel must be configured to connect to the proxy on a local port. This configuration lives in `.circleci/bazel_linux.rc` and is enabled because we overwrite the system Bazel settings in /etc/bazel.bazelrc with this file.
```
`bazel_linux.rc` --> `bazel.linux.rc`

filipesilva replied:
Done

gkalpak commented on .circleci/windows-env.ps1:
```diff
@@ -0,0 +1,30 @@
+# Install nodejs and yarn via Chocolatey.
+choco install nodejs --version 10.16.0 --no-progress
+choco install yarn --version 1.16.0 --no-progress
```
Ideally, we should use the [vendored yarn][1] (if nothing else, in order to not have one more place where we need to update yarn or risk inconsistency between linux and windows builds). But if it is too involved, maybe just leave a `TODO` comment for now.

[1]: https://github.com/angular/angular/tree/174770e6f/third_party/github.com/yarnpkg/yarn/releases

IgorMinar replied:
could we move these version numbers to env.sh so that they are not scattered all over the place?

JiaLiPassion reviewed (approved):
the zone.js part LGTM. Thanks.

IgorMinar commented on .circleci/config.yml:
```diff
@@ -669,6 +710,46 @@ jobs:
+      # Install @bazel/bazel globally and use that for the first run. Keep this version in sync
+      # with the one in package.json.
+      # Workaround for https://github.com/bazelbuild/rules_nodejs/issues/894
+      - run: yarn global add @bazel/bazel@0.27.0
```
consider extracting the version number from `package.json` in env.sh so that we don't have to update this version manually.

filipesilva replied:
Extracted to `windows-env.ps1`, reasoning in https://github.com/angular/angular/pull/31266#issuecomment-516945660.

IgorMinar commented on package.json:
```diff
@@ -26,7 +27,10 @@
-    "bazel": "bazel"
+    "bazel": "bazel",
+    "//circleci-win-comment": "See the test-win circleci job for why these are needed. If they are not needed anymore, remove them.",
+    "circleci-win-ve": "bazel test --build_tag_filters=-ivy-only --test_tag_filters=-ivy-only,-browser:chromium-local,-fixme-windows //...",
```
@alxhub could we tag only a subset of tests, primarily in compiler-cli, ngtsc, and integration tests as windows-sensitive tests and run only those for view engine and ivy? that will significantly cut down the cost of running all these tests.

IgorMinar commented on .circleci/config.yml:
```diff
@@ -669,6 +710,46 @@ jobs:
+          command: yarn circleci-win-ve
+          no_output_timeout: 120m
+      - run:
+          command: yarn circleci-win-ivy
```
shouldn't we run these tests in a separate job?

filipesilva replied:
We can do it, but it's not clear if it'll be a big improvement. The setup and Bazel download overhead might make parallelization gains marginal.

IgorMinar commented on .circleci/config.yml:
```diff
@@ -766,6 +847,16 @@ workflows:
           requires:
             - setup
+      # This job is very slow so we run it on non-PRs only for now.
+      # TODO: remove the filter when CircleCI makes Windows FS faster.
```
is there a ticket or an issue capturing this work?

filipesilva replied:
No, I was told this should get sorted before Windows on CircleCI goes on general availability.

gkalpak replied:
So, Windows on CircleCI has gone GA and things are better afaict, but still not good enough to turn on for PRs, right?
Do we expect CircleCI to continue improving speed for Windows jobs? Should we create a ticket to track it?

filipesilva replied:
I was told some improvements were in already, but I tested before/after and didn't see any difference. I've no ticket open right now.

IgorMinar commented on .circleci/windows-env.ps1:
```diff
@@ -0,0 +1,30 @@
+# Install MSYS2 packages
+C:\tools\msys64\usr\bin\bash.exe -l -c "pacman --needed --noconfirm -S zip unzip patch diffutils git"
+# Python is needed for Bazel Python targets
+choco install python --version 3.5.1 --no-progress
```
move the version to env.sh ?

IgorMinar reviewed (commented):
I left a few comments. Overall this PR is heading in the right direction, but needs more polish and turning to speed things up (both parallelization as well as filtering out tests that we don't need to run on windows).

It would also be great if we could contain all of the version numbers that we need to keep up to date in env.sh.

filipesilva commented on .circleci/config.yml:
```diff
@@ -669,6 +710,46 @@ jobs:
+      # Probably a powershell command parsing thing. This way there's no problem.
+      - run:
+          command: yarn circleci-win-ve
+          no_output_timeout: 120m
```
A couple of weeks ago the full job took ~1h30m. Now it took 2h40m. See https://angular-team.slack.com/archives/CH58NF3MG/p1564066641126100 for context.

filipesilva commented on .circleci/windows-env.ps1:
```diff
@@ -0,0 +1,30 @@
+# Install Bazel pre-reqs on Windows
+# https://docs.bazel.build/versions/master/install-windows.html
+# https://docs.bazel.build/versions/master/windows.html
+choco install msys2 --version 20180531.0.0 --no-progress --package-parameters "/NoUpdate"
```
No, this one is static. At least Bazel hasn't changed it so far from the original one they listed.

filipesilva commented on .circleci/windows-env.ps1:
```diff
@@ -0,0 +1,30 @@
+C:\tools\msys64\usr\bin\bash.exe -l -c "pacman --needed --noconfirm -S zip unzip patch diffutils git"
+# Python is needed for Bazel Python targets
+choco install python --version 3.5.1 --no-progress
+# VS Build Tools are needed for Bazel C++ targets (like com_google_protobuf)
```
I am not sure what needs com_google_protobuf, but I tried removing it and the builds failed.

filipesilva replied:
Spoke to @alexeagle, he thinks it's because of starlark docgen for `@angular/bazel` package. So we shouldn't need it for the reduced coverage.

filipesilva commented:
@IgorMinar in regardings to moving things to `env.sh`, we really should not do that. Windows does not run .sh unless they are ran from a shell like gitbash or mingw. But for the reasons I listed in the windows job defaults, we really should not use either to run Bazel commands:
```
var_6_win: &job_defaults_win
  working_directory: ~/ng
  resource_class: windows.medium
  # CircleCI windows VMs do have the GitBash shell available:
  # https://github.com/CircleCI-Public/windows-preview-docs#shells
  # But in this specific case we really should not use it because Bazel must not be ran from
  # GitBash. These issues discuss why:
  # https://github.com/bazelbuild/bazel/issues/5751
  # https://github.com/bazelbuild/bazel/issues/5724#issuecomment-410194038
  # https://github.com/bazelbuild/bazel/issues/6339#issuecomment-441600879
  shell: powershell.exe -ExecutionPolicy Bypass
  machine:
    image: windows-server-2019
```

Even if we could use gitbash/mingw we probably shouldn't because the goal is to test things work on windows and not on the mock unix environment provided by those shells. There are especially problematic cases related to [path conversions](https://stackoverflow.com/questions/7250130/how-to-stop-mingw-and-msys-from-mangling-path-names-given-at-the-command-line).

So even though we have an `env.sh` script where we want to put versions, we can't use it because it's not cross platform. The second best place is `.circleci/windows-env.ps1` I think.

filipesilva commented:
Split the job into a VE and an Ivy one, made them depend on the correspondent linux job, and limited tests to `//packages/compiler-cli/...`. Each takes around 35m now. Also refactored the anchors a bit to better match the linux ones. It would still be the longest job on PRs so I don't think we should turn it on for them.

I'd really like to use CircleCI executors and commands (like [we do in the CLI](https://github.com/angular/angular-cli/blob/01e9c673fa2209a5aa0b87b7b3fe0524a3d9f922/.circleci/config.yml#L28-L108) atm), but that requires using 2.1 configurations and CircleCI pipelines to be turned. Last time I tried to turn those on that broke some jobs because of the `CIRCLE_COMPARE_URL` disappeared and we rely on it for non-PR jobs. Still waiting to hear if that's fixed on CircleCI's side.

josephperrott commented on .circleci/bazel.common.rc:
```diff
@@ -1,28 +1,11 @@
-# speeding up the analysis time significantly with Bazel managed node dependencies on the CI.
-build --repository_cache=/home/circleci/bazel_repository_cache
-
 # Don't be spammy in the logs
```
Should we include a note at the top reminding that all rules in this rc file should be OS agnostic?

filipesilva replied:
Yes that would be better. Added.

josephperrott commented on .circleci/config.yml:
```diff
@@ -27,6 +27,7 @@ var_2: &browsers_docker_image circleci/node:10.16-browsers
 # **NOTE 2 **: Keep the static part of the cache key as prefix to enable correct fallbacks.
 # See https://circleci.com/docs/2.0/caching/#restoring-cache for how prefixes work in CircleCI.
 var_3: &cache_key v3-angular-node-10.16-{{ checksum "yarn.lock" }}-{{ checksum "WORKSPACE" }}-{{ checksum "packages/bazel/package.bzl" }}-{{ checksum "aio/yarn.lock" }}
+var_3: &cache_key_win v3-angular-win-node-10.16-{{ checksum "yarn.lock" }}-{{ checksum "WORKSPACE" }}-{{ checksum "packages/bazel/package.bzl" }}-{{ checksum "aio/yarn.lock" }}
```
Should this be `var_3_win`?

filipesilva replied:
Indeed it should, fixed now.

josephperrott commented on .circleci/config.yml:
```diff
@@ -669,6 +719,54 @@ jobs:
+      # Install @bazel/bazel globally and use that for the first run.
+      # Workaround for https://github.com/bazelbuild/rules_nodejs/issues/894
+      - run: yarn global add @bazel/bazel@$env:BAZEL_VERSION
+      - run: bazel info
```
Do we still need to run `bazel info`?

filipesilva replied:
AS far as I can tell, yes. I tried taking it off and it didn't work anymore, I got the error in https://github.com/bazelbuild/rules_nodejs/issues/894 back. I don't quite understand why, I thought `bazel info` didn't run the install rule. But maybe it does?

IgorMinar commented:
.circleci/windows-env.ps1 sounds good. thanks

josephperrott commented on .circleci/config.yml:
```diff
@@ -669,6 +724,54 @@ jobs:
              yarn --cwd packages/zone.js electrontest
 
+  # Windows jobs
+  # CircleCI support for Windows jobs is still in preview.
```
Windows CI has moved to GA, should we link to https://circleci.com/docs/2.0/hello-world-windows/ instead?

filipesilva replied:
Yes, we should. Fixed.

josephperrott reviewed (commented):
Just one last comment, and then I think its good to go

filipesilva commented:
There seems to be some problem with cache restoration:
- without cache: https://circleci.com/gh/angular/angular/416574
- with cache: https://circleci.com/gh/angular/angular/416679

gkalpak commented on .circleci/config.yml:
```diff
@@ -65,6 +74,24 @@ var_6: &job_defaults
+    # - Node 12
+    # - yarn 1.17
+    # - Python 3 3.7.4
+    image: windows-server-2019-vs2019:201908-02
```
A comment in `&init_environment_win` says _"need to install node and yarn before, as the base windows image doesn't have anything"_. Is that still needed?

filipesilva replied:
Good catch, changed that comment.

gkalpak commented on .circleci/config.yml:
```diff
@@ -766,6 +868,22 @@ workflows:
+      # TODO: remove the filter when CircleCI makes Windows FS faster.
+      - test_win:
+          requires:
+            - test
```
Wouldn't that make this job (and thus the overall CI run) ~10mins slower?
I assume it it to save resources, but is it worth it (especially given that this is not run on PRs, so theoretically all non-Windows tests should pass most of the time)?

filipesilva replied:
I think the 10m don't make a difference on master, and that removing it now is risking forgetting to turn it back on when it goes on PRs. But I am assuming no one is waiting and waiting for master to finish runs.

gkalpak replied:
Fair enough :smiley:
I think the caretaker _is_ waiting when they want to cut a release, but maybe 10m isn't a big deal for that usecase.

gkalpak replied:
I would stay it is still worth adding a comment on why we have this job dependency (not necessary, makes build slower, saves resources).

filipesilva replied:
Done

gkalpak reviewed (approved):
🤞

filipesilva commented:
caretaker: the CI failures on `test_aio`, `test_aio_local` and `test_aio_local_ivy` can be ignored. Those tests will always fail from non-fork PRs like this one.
